File: lint-arwaky/archive/capabilities/mandatory_inheritance_checker.py
"""mandatory_inheritance_checker — AES027: mandatory contract inheritance for agent/capabilities/infrastructure."""

from __future__ import annotations

import logging

from ..taxonomy import (
    AdapterName,
    ColumnNumber,
    ErrorCode,
    FilePath,
    FilePathList,
    LineNumber,
    LintMessage,
    LintResult,
    LintResultList,
    Severity,
    Identity,
)
# contract imports removed — not used in this module

logger = logging.getLogger(__name__)

# Map layer name → contract suffix to check inheritance against
LAYER_CONTRACT_SUFFIX = {
    "infrastructure": "_port",
    "capabilities": "_protocol",
    "agent": "_aggregate",
}


class MandatoryInheritanceChecker:
    """Check that files in agent/capabilities/infrastructure that import
    at least one contract symbol also have a class inheriting from it.

    Surfaces and contract files import contracts for composition/call —
    they are NOT checked by this rule.
    """

    @property
    def rule_name(self) -> Identity:
        return Identity(value="mandatory_inheritance")

    def check_mandatory_inheritance(
        self,
        analyzer,
        files: FilePathList,
        root_dir: FilePath,
        results: LintResultList,
    ) -> None:
        """Verify 1:1 mapping: if a file in agent/capabilities/infrastructure
        imports a contract, it must have a class inheriting from at least one of them.
        """
        for f in files:
            layer_vo = analyzer._detect_layer(f, root_dir)
            if not layer_vo:
                continue

            layer_str = str(layer_vo).split("(")[0]
            if layer_str not in LAYER_CONTRACT_SUFFIX:
                continue  # only agent/capabilities/infrastructure

            # __init__.py barrel files don't need to inherit
            if f.value.endswith("__init__.py"):
                # But barrel files must have at least one class inheriting
                # from a contract (imported and re-exported)
                continue

            # Get imports
            symbols = analyzer.parser.get_raw_symbols(f)
            imports = symbols.value.get("imports", {})
            contract_imports = self._filter_contract_imports(imports)
            if not contract_imports:
                continue  # no contract imported → skip, surface-like

            # Get class definitions
            metadata = analyzer.parser.get_class_definitions(f)
            classes = metadata.value.get("classes", [])
            resolved_bases = set()
            for cls in classes:
                if not isinstance(cls, dict):
                    continue
                for base in cls.get("resolved_bases") or cls.get("bases") or []:
                    resolved_bases.add(str(base))

            # Check: does any resolved base name match a contract import?
            inherited = False
            for contract_name in contract_imports:
                if contract_name in resolved_bases:
                    inherited = True
                    break

            if not inherited:
                imported_list = ", ".join(sorted(contract_imports))
                results.append(
                    LintResult(
                        file=f,
                        line=LineNumber(value=0),
                        column=ColumnNumber(value=0),
                        code=ErrorCode(code="AES027"),
                        message=LintMessage(
                            value=(
                                f"AES027 MANDATORY_INHERITANCE_VIOLATION: File imports contracts "
                                f"({imported_list}) but no class inherits from any of them. "
                                f"Layer '{layer_str}' must implement its contract via inheritance. "
                                f"FIX: Make at least one class in this file inherit from one of the imported contracts."
                            )
                        ),
                        source=AdapterName(value="architecture"),
                        severity=Severity.CRITICAL,
                    )
                )

    def _filter_contract_imports(self, imports: dict) -> set[str]:
        """Filter import names that refer to contract layer symbols
        (suffixes: _port, _protocol, _aggregate, or names starting with 'I').

        Returns set of symbol names that are contract imports.
        """
        contract_suffixes = ("_port", "_protocol", "_aggregate")
        result = set()
        for imp_name, imp_info in imports.items():
            # Check imported symbol name
            if imp_name.startswith("I") and any(
                imp_name.endswith(s) for s in contract_suffixes
            ):
                result.add(imp_name)
                continue
            # Check module path
            module = str(imp_info.get("module", "")) if isinstance(imp_info, dict) else str(imp_info)
            if "contract" in module.lower() and any(
                imp_name.endswith(s) for s in contract_suffixes
            ):
                result.add(imp_name)
            elif any(imp_name.endswith(s) for s in contract_suffixes):
                result.add(imp_name)
            # Interface names (I-prefixed from contract) — they end up as resolved names
            elif imp_name.startswith("I"):
                result.add(imp_name)

        return result

File: lint-arwaky/src-rust/capabilities/mandatory_inheritance_checker.rs
// mandatory_inheritance_checker — AES027: mandatory contract inheritance.
// Checks that files in agent/capabilities/infrastructure that import contracts
// also have a class inheriting from them.

use std::fs;
use regex::Regex;
use crate::taxonomy::{
    AdapterName, ColumnNumber, ErrorCode, FilePath, LayerNameVO,
    LintMessage, LintResult, LineNumber, Severity, ScopeRef, LocationList,
    ArchitectureConfig,
};

// Map layer name → required contract suffix pattern
const LAYER_CONTRACT_SUFFIX: &[(&str, &str)] = &[
    ("infrastructure", "_port"),
    ("capabilities", "_protocol"),
    ("agent", "_aggregate"),
];

/// Check that files in agent/capabilities/infrastructure that import contracts
/// also have a class inheriting from them (AES027).
pub struct MandatoryInheritanceChecker {
    config: ArchitectureConfig,
}

impl MandatoryInheritanceChecker {
    pub fn new(config: ArchitectureConfig) -> Self {
        Self { config }
    }

    fn make_result(file: &str, msg: &str) -> LintResult {
        LintResult {
            file: FilePath::new(file.to_string()),
            line: LineNumber::new(0),
            column: ColumnNumber::new(0),
            code: ErrorCode::new("AES027"),
            message: LintMessage::new(msg),
            source: AdapterName::new("architecture"),
            severity: Severity::CRITICAL,
            enclosing_scope: ScopeRef {
                name: "".to_string(),
                kind: "".to_string(),
                file: FilePath::new(""),
                start_line: LineNumber::new(0),
                end_line: LineNumber::new(0),
            },
            related_locations: LocationList::new(Vec::new()),
        }
    }

    fn detect_file_layer(&self, file: &str, root_dir: &str) -> Option<String> {
        let rel = file.strip_prefix(root_dir).unwrap_or(file).trim_start_matches('/');

        let mut layers: Vec<(&LayerNameVO, &crate::taxonomy::LayerDefinition)> =
            self.config.layers.iter().collect();
        layers.sort_by(|a, b| b.1.path.value.len().cmp(&a.1.path.value.len()));

        for (name, def) in layers {
            if rel.starts_with(def.path.value.as_str()) {
                return Some(name.value.clone());
            }
        }
        None
    }

    fn is_contract_import(name: &str, module: &str) -> bool {
        let contract_suffixes = ["_port", "_protocol", "_aggregate"];
        let has_suffix = contract_suffixes.iter().any(|s| name.ends_with(s));
        let is_interface = name.starts_with('I') && name.len() > 1;
        let from_contract = module.contains("contract");

        (has_suffix || (is_interface && has_suffix)) || (from_contract && has_suffix)
    }

    fn extract_contract_imports(content: &str) -> Vec<String> {
        let mut imports: Vec<String> = Vec::new();
        // Match `from x import Y, Z` or `from x import (Y, Z)`
        let from_re = Regex::new(r"from\s+(\S+)\s+import\s+(.+)").unwrap();

        for line in content.lines() {
            let trimmed = line.trim();
            if let Some(caps) = from_re.captures(trimmed) {
                let module = caps.get(1).map(|m| m.as_str()).unwrap_or("");
                let names_str = caps.get(2).map(|m| m.as_str()).unwrap_or("");
                // Parse imported names (handle parentheses on single line)
                let names = names_str
                    .trim_matches(|c| c == '(' || c == ')')
                    .split(',')
                    .map(|s| s.trim().split_whitespace().next().unwrap_or("").to_string())
                    .filter(|s| !s.is_empty());

                for name in names {
                    if Self::is_contract_import(&name, module) {
                        imports.push(name);
                    }
                }
            }
        }
        imports
    }

    fn extract_class_bases(content: &str) -> Vec<String> {
        let class_re = Regex::new(r"class\s+\w+\s*\(([^)]*)\)").unwrap();
        let mut bases: Vec<String> = Vec::new();

        for line in content.lines() {
            if let Some(caps) = class_re.captures(line) {
                let base_str = caps.get(1).map(|m| m.as_str()).unwrap_or("");
                for base in base_str.split(',') {
                    let base_name = base.trim().to_string();
                    if !base_name.is_empty() {
                        bases.push(base_name);
                    }
                }
            }
        }
        bases
    }

    /// Check mandatory inheritance for a list of files.
    pub fn check_mandatory_inheritance(
        &self,
        files: &[String],
        root_dir: &str,
        violations: &mut Vec<LintResult>,
    ) {
        for file in files {
            let basename = file.split('/').last().unwrap_or("");

            // Skip barrel files
            if basename == "__init__.py" || basename == "mod.rs" {
                continue;
            }

            let layer = match self.detect_file_layer(file, root_dir) {
                Some(l) => l,
                None => continue,
            };

            // Only check agent/capabilities/infrastructure layers
            let layer_suffix = match LAYER_CONTRACT_SUFFIX.iter().find(|(l, _)| *l == layer.as_str()) {
                Some((_, s)) => *s,
                None => continue,
            };

            let Ok(content) = fs::read_to_string(file) else { continue; };

            let contract_imports = Self::extract_contract_imports(&content);
            if contract_imports.is_empty() {
                continue; // No contract imported → surface-like, skip
            }

            let class_bases = Self::extract_class_bases(&content);

            // Check: does any class base match a contract import?
            let inherited = contract_imports.iter().any(|ci|
                class_bases.iter().any(|base| base.contains(ci.as_str()))
            );

            if !inherited {
                let imported_list = contract_imports.join(", ");
                let msg = format!(
                    "AES027 MANDATORY_INHERITANCE_VIOLATION: File imports contracts ({}) but no class inherits from any of them. \
                    Layer '{}' must implement its contract via inheritance. \
                    FIX: Make at least one class in this file inherit from one of the imported contracts.",
                    imported_list, layer
                );
                violations.push(Self::make_result(file, &msg));
            }
        }
    }
}
